cmd/tools/test/find_altered_tests.go

```diff
@@ -0,0 +1,188 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+// find_altered_tests.go
+//
+// This program identifies altered or added test suite names based on the provided categories.
+// It accepts the following inputs via command-line flags:
+//
+// - Category (-c): The category of tests to find (e.g., unit, integration, functional, functional_ndc).
+// - Source Git Reference (-s): The source Git reference (commit SHA, branch, etc.).
+// - Target Git Reference (-t): The target Git reference (commit SHA, branch, etc.).
+//
+// The program outputs environment variable assignments in the format `key=value`,
+// where each key corresponds to a modified test suite category, and the value is a
+// pipe-separated list of altered test suites.
+//
+// Usage:
+//   go run find_altered_tests.go -c <category1> -c <category2> ... -s <sourceRef> -t <targetRef>
+
+// CategoryDirs maps test categories to their corresponding directories
+var CategoryDirs = map[string][]string{
+	"unit":           {"./client", "./common", "./internal", "./service", "./temporal", "./tools", "./cmd"},
+	"integration":    {"./common/persistence/tests", "./tools/tests", "./temporaltest", "./internal/temporalite"},
+	"functional":     {"./tests"},
+	"functional_ndc": {"./tests/ndc"},
+	"functional_xdc": {"./tests/xdc"},
+}
```

Comment:
I can almost guarantee this will get out of sync with the Makefile. Let's have one source of truth. I would recommend adding a make target to the Makefile that will inject all of these directories into the script via args.

I would have preferred having the source of truth in Go and not having a Makefile at all but my recommendation fits better with today's reality.
An alternative would be to put a comment in both locations to prompt users to update the other location when something changes but that's probably overkill.

Reply:
As discussed, will add comment for the time being